### PR TITLE
Initial development setup

### LINKED ISSUE #1: Initial node setup
We need to the initial repository setup:

1. editorconfig + initial readme + initial package.json
2. eslint
3. babel 
4. jest
5. travis

Side note: I recommend tabs [for accessibility reasons](https://twitter.com/sarah_federman/status/1146544481556033537?lang=en) but I believe that spaces are preferred at an org level, so we'll go with spaces!

### PR DESCRIPTION
**Pull request by slifty**
## Description
This PR adds and configures several core developer resources for this repository:

- community files
- eslint (and .editorconfig)
- jest
- babel
- travis

This PR does more than I would normally add in a single PR, but since it is initial setup I figured the full context is valuable.

Resolves #1

## Steps to Test
- `npm install`
- `npm run lint`
- `npm test`
- `npm run build`

### CODE REVIEW

**Comment by slifty** [comment]
I don't want to disrupt everybody by assigning reviews to the whole team, but please feel free to weigh in / review as well @xmunoz and @andrewatwood!

File: README.md [hunk 1]
```diff
@@ -1,2 +1,39 @@
-# upload-service
-A service responsible for handling upload requests
+# Permanent.org Upload Service
+
+This repository contins a service responsible for handling upload requests.
```
Comment (jasonaowen) [inline README.md:3]:
```suggestion
This repository contains a service responsible for handling upload requests.
```

File: .editorconfig [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+[*]
+charset = utf-8
+end_of_line = lf
+indent_size = 2
+indent_style = space
+insert_final_newline = true
+trim_trailing_whitespace = true
+
+[*.{md,mdx}]
+trim_trailing_whitespace = false
```
Comment (jasonaowen) [inline .editorconfig:12]: Interesting, why explicitly allow trailing whitespace in markdown files?

File: package.json [hunk 3]
```diff
@@ -3,7 +3,9 @@
   "version": "0.0.0",
   "description": "A service responsible for handling upload requests",
   "main": "lib/index.js",
-  "scripts": {},
+  "scripts": {
+    "lint": "eslint 'src/**/*.js'"
```
Comment (jasonaowen) [inline package.json]: We haven't discussed this yet, but we've started to use TypeScript more in the front end, and I'd like to use it here as well!

**Review by jasonaowen — APPROVED** [review]
Thank you for the well-written and carefully structured commits! They make it super easy to review.

I'm generally +1 to this (after typo fix, below), but as a follow-up I'd love to make the case for using TypeScript. We're already using it in the front end, it prevents or catches a whole category of programming errors, and the beginning of a repo is the best time to adopt it. If that's something y'all are happy with, I'm fine with that being in this PR or in a follow-up PR, and with @slifty  or I doing that work.

I also want to explicitly +1 some things in the commit messages: using the current active LTS release of Node.js, using npm instead of yarn, using Airbnb's eslint config, using babel (of course), using Jest, and using Travis CI.

File: package.json (hunk 3 above)
Comment (andrewatwood) [reply package.json]: +1 to TypeScript, the boilerplate and config can be a little more persnickety but once it's all set up I love it on the backend too.

File: .editorconfig (hunk 2 above)
Comment (slifty) [reply .editorconfig:12]: Great question and I'm gonna remove this (I copied from another project and missed this when reviewing)

File: package.json (hunk 3 above)
Comment (slifty) [reply package.json]: I have not used TS in the wild and would love to!  Would you be up for adding that in after I merge this?

Comment (jasonaowen) [reply package.json]: Sure!

Comment (slifty) [reply package.json]: Update: I'm going to take a stab at putting it in this PR

File: .travis.yml [hunk 4]
```diff
@@ -0,0 +1,8 @@
+language: node_js
+
+script:
+  - npm run lint
+  - npm run test
+
+node_js:
+  - 12
```
Comment (xmunoz) [inline .travis.yml:8]: If we are 100% sure that this will run as a microservice, than 12 is fine. Otherwise, use 10.

**Review by xmunoz — CHANGES_REQUESTED** [review]

File: .travis.yml (hunk 4 above)
Comment (jasonaowen) [reply .travis.yml:8]: I think part of the goal here is to bind our hands to ensure that we will not deploy it alongside the existing stack!

**Comment by slifty** [comment]
Typescript has been added, and feedback has been addressed.  Ready for a re-review @jasonaowen!

**Review by jvasile — APPROVED** [review]
License stuff is good to go.

**Review by xmunoz — APPROVED** [review]

File: .eslintrc.json [hunk 5]
```diff
@@ -1,5 +1,14 @@
...
-  "extends": "airbnb-base",
+  "parser": "@typescript-eslint/parser",
+  "parserOptions": {
+    "project": "./tsconfig.json"
+  },
+  "plugins": ["@typescript-eslint"],
+  "extends": [
+    "airbnb-base",
+    "plugin:@typescript-eslint/recommended",
+    "plugin:@typescript-eslint/recommended-requiring-type-checking"
```
Comment (jasonaowen) [inline .eslintrc.json:10]:
```suggestion
    "plugin:@typescript-eslint/all",
    "airbnb-typescript/base"
```

Note that the order does have an effect; I believe the later entries override the earlier entries, which is what I think we want here, but I haven't been able to find any documentation on that and haven't dived into the eslint source code.

**Review by jasonaowen — APPROVED** [review]
@slifty, you mentioned to me that you were surprised that the linter wasn't picking up some things you expected it to. I tried it out, and I agree! In particular, the config as-is allows the following legal but terrible TypeScript:

```
type example = { 
  foo: boolean;
  bar:string,
};
```

It turns out there is a stricter ruleset, [`@typescript-eslint/all`](https://github.com/typescript-eslint/typescript-eslint/blob/v2.0.0/packages/eslint-plugin/src/configs/all.json). Using that turned up more of the errors I expected.

The other wrinkle here is that the `airbnb-base` configuration does not include any TypeScript rules, and there are several rules provided by `@typescript-eslint` which shadow base ESLint rules so that they can take into account differences in the TypeScript AST. There is another package, [`eslint-config-airbnb-typescript`](https://www.npmjs.com/package/eslint-config-airbnb-typescript), which disables those airbnb rules and replaces them with their TypeScript variants. Unfortunately, it introduces some spurious peer dependency warnings, which we can ignore because we're not working with react; that's a known issue.

Once I installed the additional package and made the change suggested below, I got the linter errors I was hoping for on the above code:

```
  1:6   error  Type Alias name `example` must match one of the following formats: PascalCase  @typescript-eslint/naming-convention
  1:6   error  Use an `interface` instead of a `type`                                         @typescript-eslint/consistent-type-definitions
  1:6   error  'example' is defined but never used                                            @typescript-eslint/no-unused-vars
  1:16  error  Type literals are not allowed                                                  @typescript-eslint/no-type-alias
  3:6   error  Expected a space after the ':'                                                 @typescript-eslint/type-annotation-spacing
  3:14  error  Expected a semicolon                                                           @typescript-eslint/member-delimiter-style
```

Finally, I'll add that while this proposed configuration is fairly strict, I'm totally open to disabling rules we discuss and disagree with. I'd rather start too-strict and loosen up later than start too loose and try to implement additional rules later.

So, overall +1 with suggested changes. If you agree with and are able to make these changes, please go ahead and merge after doing so; otherwise I'm happy to follow up with a separate pull request.

**Pull request merged**